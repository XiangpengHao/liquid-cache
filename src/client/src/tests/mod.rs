use std::{path::Path, sync::Arc};

use datafusion::{
    physical_plan::{ExecutionPlan, display::DisplayableExecutionPlan},
    prelude::{SessionConfig, SessionContext},
};

use crate::LiquidCacheBuilder;

async fn setup_tpch_ctx() -> Arc<SessionContext> {
    let mut session_config = SessionConfig::from_env().unwrap();
    // To get deterministic results
    session_config.options_mut().execution.target_partitions = 8;
    let ctx = LiquidCacheBuilder::new("http://localhost:50051")
        .build(session_config)
        .unwrap();
    let tables = [
        "customer", "lineitem", "nation", "orders", "part", "partsupp", "region", "supplier",
    ];

    for table in tables {
        let table_url = format!("../../benchmark/tpch/data/sf0.001/{}.parquet", table);
        ctx.register_parquet(table, table_url, Default::default())
            .await
            .unwrap();
    }
    Arc::new(ctx)
}

/// One query file can contain multiple queries, separated by `;`
fn get_query_by_id(query_dir: impl AsRef<Path>, query_id: u32) -> Vec<String> {
    let query_dir = query_dir.as_ref();
    let mut path = query_dir.to_owned();
    path.push(format!("q{query_id}.sql"));
    let content = std::fs::read_to_string(&path).unwrap();
    content
        .split(';')
        .map(|s| s.trim())
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string())
        .collect()
}

async fn get_query_plan_by_id(query_id: u32) -> Arc<dyn ExecutionPlan> {
    let query_dir = Path::new("../../benchmark/tpch/queries");
    let ctx = setup_tpch_ctx().await;

    let query = if query_id != 15 {
        &get_query_by_id(query_dir, query_id)[0]
    } else {
        &get_query_by_id(query_dir, query_id)[1]
    };
    let result = ctx.sql(query).await.unwrap();
    let (state, plan) = result.into_parts();
    state.create_physical_plan(&plan).await.unwrap()
}

#[tokio::test]
async fn test_tpch_q1() {
    let physical_plan = get_query_plan_by_id(1).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q2() {
    let physical_plan = get_query_plan_by_id(2).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q3() {
    let physical_plan = get_query_plan_by_id(3).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q4() {
    let physical_plan = get_query_plan_by_id(4).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q5() {
    let physical_plan = get_query_plan_by_id(5).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q6() {
    let physical_plan = get_query_plan_by_id(6).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q7() {
    let physical_plan = get_query_plan_by_id(7).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q8() {
    let physical_plan = get_query_plan_by_id(8).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q9() {
    let physical_plan = get_query_plan_by_id(9).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q10() {
    let physical_plan = get_query_plan_by_id(10).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q11() {
    let physical_plan = get_query_plan_by_id(11).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q12() {
    let physical_plan = get_query_plan_by_id(12).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q13() {
    let physical_plan = get_query_plan_by_id(13).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q14() {
    let physical_plan = get_query_plan_by_id(14).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
/// Q15 needs to actually run the first query to get the correct plan
#[ignore]
async fn test_tpch_q15() {
    let physical_plan = get_query_plan_by_id(15).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q16() {
    let physical_plan = get_query_plan_by_id(16).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q17() {
    let physical_plan = get_query_plan_by_id(17).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q18() {
    let physical_plan = get_query_plan_by_id(18).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q19() {
    let physical_plan = get_query_plan_by_id(19).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q20() {
    let physical_plan = get_query_plan_by_id(20).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q21() {
    let physical_plan = get_query_plan_by_id(21).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}

#[tokio::test]
async fn test_tpch_q22() {
    let physical_plan = get_query_plan_by_id(22).await;
    let displayable = DisplayableExecutionPlan::new(physical_plan.as_ref());
    insta::assert_snapshot!(displayable.tree_render().to_string());
}
