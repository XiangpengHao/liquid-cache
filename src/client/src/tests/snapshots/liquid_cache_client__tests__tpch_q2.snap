---
source: src/client/src/tests/mod.rs
expression: displayable.tree_render().to_string()
---
┌───────────────────────────┐
│  SortPreservingMergeExec  │
│    --------------------   │
│ s_acctbal DESC, n_name ASC│
│   NULLS LAST, s_name ASC  │
│    NULLS LAST, p_partkey  │
│       ASC NULLS LAST      │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│          SortExec         │
│    --------------------   │
│ s_acctbal@0 DESC, n_name@2│
│  ASC NULLS LAST, s_name@1 │
│       ASC NULLS LAST,     │
│       p_partkey@3 ASC     │
│         NULLS LAST        │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│       ProjectionExec      │
│    --------------------   │
│       n_name: n_name      │
│       p_mfgr: p_mfgr      │
│    p_partkey: p_partkey   │
│    s_acctbal: s_acctbal   │
│    s_address: s_address   │
│    s_comment: s_comment   │
│       s_name: s_name      │
│      s_phone: s_phone     │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│    CoalesceBatchesExec    │
│    --------------------   │
│     target_batch_size:    │
│           16384           │
└─────────────┬─────────────┘
┌─────────────┴─────────────┐
│        HashJoinExec       │
│    --------------------   │
│            on:            │
│ (p_partkey = ps_partkey), │
│    (ps_supplycost = min   ├──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│         (partsupp         │                                                                                                                                  │
│       .ps_supplycost      │                                                                                                                                  │
│             ))            │                                                                                                                                  │
└─────────────┬─────────────┘                                                                                                                                  │
┌─────────────┴─────────────┐                                                                                                                    ┌─────────────┴─────────────┐
│   CoalescePartitionsExec  │                                                                                                                    │       ProjectionExec      │
│                           │                                                                                                                    │    --------------------   │
│                           │                                                                                                                    │ min(partsupp.ps_supplycost│
│                           │                                                                                                                    │             ):            │
│                           │                                                                                                                    │min(partsupp.ps_supplycost)│
│                           │                                                                                                                    │                           │
│                           │                                                                                                                    │        ps_partkey:        │
│                           │                                                                                                                    │         ps_partkey        │
└─────────────┬─────────────┘                                                                                                                    └─────────────┬─────────────┘
┌─────────────┴─────────────┐                                                                                                                    ┌─────────────┴─────────────┐
│    CoalesceBatchesExec    │                                                                                                                    │       AggregateExec       │
│    --------------------   │                                                                                                                    │    --------------------   │
│     target_batch_size:    │                                                                                                                    │           aggr:           │
│           16384           │                                                                                                                    │min(partsupp.ps_supplycost)│
│                           │                                                                                                                    │                           │
│                           │                                                                                                                    │    group_by: ps_partkey   │
│                           │                                                                                                                    │                           │
│                           │                                                                                                                    │           mode:           │
│                           │                                                                                                                    │      FinalPartitioned     │
└─────────────┬─────────────┘                                                                                                                    └─────────────┬─────────────┘
┌─────────────┴─────────────┐                                                                                                                    ┌─────────────┴─────────────┐
│        HashJoinExec       │                                                                                                                    │    CoalesceBatchesExec    │
│    --------------------   │                                                                                                                    │    --------------------   │
│            on:            ├──────────────┐                                                                                                     │     target_batch_size:    │
│(r_regionkey = n_regionkey)│              │                                                                                                     │           16384           │
└─────────────┬─────────────┘              │                                                                                                     └─────────────┬─────────────┘
┌─────────────┴─────────────┐┌─────────────┴─────────────┐                                                                                       ┌─────────────┴─────────────┐
│   LiquidCacheClientExec   ││       ProjectionExec      │                                                                                       │      RepartitionExec      │
│    --------------------   ││    --------------------   │                                                                                       │    --------------------   │
│          server:          ││       n_name: n_name      │                                                                                       │ partition_count(in->out): │
│  http://localhost:50051,  ││                           │                                                                                       │           8 -> 8          │
│        mode=liquid,       ││        n_regionkey:       │                                                                                       │                           │
│      object_stores=[]     ││        n_regionkey        │                                                                                       │    partitioning_scheme:   │
│                           ││                           │                                                                                       │  Hash([ps_partkey@0], 8)  │
│                           ││       p_mfgr: p_mfgr      │                                                                                       │                           │
│                           ││    p_partkey: p_partkey   │                                                                                       │                           │
│                           ││                           │                                                                                       │                           │
│                           ││       ps_supplycost:      │                                                                                       │                           │
│                           ││       ps_supplycost       │                                                                                       │                           │
│                           ││                           │                                                                                       │                           │
│                           ││    s_acctbal: s_acctbal   │                                                                                       │                           │
│                           ││    s_address: s_address   │                                                                                       │                           │
│                           ││    s_comment: s_comment   │                                                                                       │                           │
│                           ││       s_name: s_name      │                                                                                       │                           │
│                           ││      s_phone: s_phone     │                                                                                       │                           │
└─────────────┬─────────────┘└─────────────┬─────────────┘                                                                                       └─────────────┬─────────────┘
┌─────────────┴─────────────┐┌─────────────┴─────────────┐                                                                                       ┌─────────────┴─────────────┐
│       DataSourceExec      ││    CoalesceBatchesExec    │                                                                                       │       AggregateExec       │
│    --------------------   ││    --------------------   │                                                                                       │    --------------------   │
│          files: 1         ││     target_batch_size:    │                                                                                       │           aggr:           │
│      format: parquet      ││           16384           │                                                                                       │min(partsupp.ps_supplycost)│
│                           ││                           │                                                                                       │                           │
│         predicate:        ││                           │                                                                                       │    group_by: ps_partkey   │
│      r_name = EUROPE      ││                           │                                                                                       │       mode: Partial       │
└───────────────────────────┘└─────────────┬─────────────┘                                                                                       └─────────────┬─────────────┘
                             ┌─────────────┴─────────────┐                                                                                       ┌─────────────┴─────────────┐
                             │        HashJoinExec       │                                                                                       │    CoalesceBatchesExec    │
                             │    --------------------   │                                                                                       │    --------------------   │
                             │            on:            ├──────────────┐                                                                        │     target_batch_size:    │
                             │(n_nationkey = s_nationkey)│              │                                                                        │           16384           │
                             └─────────────┬─────────────┘              │                                                                        └─────────────┬─────────────┘
                             ┌─────────────┴─────────────┐┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐
                             │   LiquidCacheClientExec   ││       ProjectionExec      │                                                          │        HashJoinExec       │
                             │    --------------------   ││    --------------------   │                                                          │    --------------------   │
                             │          server:          ││       p_mfgr: p_mfgr      │                                                          │            on:            │
                             │  http://localhost:50051,  ││    p_partkey: p_partkey   │                                                          │(r_regionkey = n_regionkey)│
                             │        mode=liquid,       ││                           │                                                          │                           │
                             │      object_stores=[]     ││       ps_supplycost:      │                                                          │                           │
                             │                           ││       ps_supplycost       │                                                          │                           │
                             │                           ││                           │                                                          │                           │
                             │                           ││    s_acctbal: s_acctbal   │                                                          │                           ├──────────────┐
                             │                           ││    s_address: s_address   │                                                          │                           │              │
                             │                           ││    s_comment: s_comment   │                                                          │                           │              │
                             │                           ││       s_name: s_name      │                                                          │                           │              │
                             │                           ││                           │                                                          │                           │              │
                             │                           ││        s_nationkey:       │                                                          │                           │              │
                             │                           ││        s_nationkey        │                                                          │                           │              │
                             │                           ││                           │                                                          │                           │              │
                             │                           ││      s_phone: s_phone     │                                                          │                           │              │
                             └─────────────┬─────────────┘└─────────────┬─────────────┘                                                          └─────────────┬─────────────┘              │
                             ┌─────────────┴─────────────┐┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                             │       DataSourceExec      ││    CoalesceBatchesExec    │                                                          │   LiquidCacheClientExec   ││       ProjectionExec      │
                             │    --------------------   ││    --------------------   │                                                          │    --------------------   ││    --------------------   │
                             │          files: 1         ││     target_batch_size:    │                                                          │          server:          ││        n_regionkey:       │
                             │      format: parquet      ││           16384           │                                                          │  http://localhost:50051,  ││        n_regionkey        │
                             │                           ││                           │                                                          │        mode=liquid,       ││                           │
                             │                           ││                           │                                                          │      object_stores=[]     ││        ps_partkey:        │
                             │                           ││                           │                                                          │                           ││         ps_partkey        │
                             │                           ││                           │                                                          │                           ││                           │
                             │                           ││                           │                                                          │                           ││       ps_supplycost:      │
                             │                           ││                           │                                                          │                           ││       ps_supplycost       │
                             └───────────────────────────┘└─────────────┬─────────────┘                                                          └─────────────┬─────────────┘└─────────────┬─────────────┘
                                                          ┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                                                          │        HashJoinExec       │                                                          │       DataSourceExec      ││    CoalesceBatchesExec    │
                                                          │    --------------------   │                                                          │    --------------------   ││    --------------------   │
                                                          │            on:            │                                                          │          files: 1         ││     target_batch_size:    │
                                                          │  (s_suppkey = ps_suppkey) ├──────────────┐                                           │      format: parquet      ││           16384           │
                                                          │                           │              │                                           │                           ││                           │
                                                          │                           │              │                                           │         predicate:        ││                           │
                                                          │                           │              │                                           │      r_name = EUROPE      ││                           │
                                                          └─────────────┬─────────────┘              │                                           └───────────────────────────┘└─────────────┬─────────────┘
                                                          ┌─────────────┴─────────────┐┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐
                                                          │   LiquidCacheClientExec   ││      RepartitionExec      │                                                          │        HashJoinExec       │
                                                          │    --------------------   ││    --------------------   │                                                          │    --------------------   │
                                                          │          server:          ││ partition_count(in->out): │                                                          │            on:            │
                                                          │  http://localhost:50051,  ││           1 -> 8          │                                                          │(n_nationkey = s_nationkey)├──────────────┐
                                                          │        mode=liquid,       ││                           │                                                          │                           │              │
                                                          │      object_stores=[]     ││    partitioning_scheme:   │                                                          │                           │              │
                                                          │                           ││     RoundRobinBatch(8)    │                                                          │                           │              │
                                                          └─────────────┬─────────────┘└─────────────┬─────────────┘                                                          └─────────────┬─────────────┘              │
                                                          ┌─────────────┴─────────────┐┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                                                          │       DataSourceExec      ││    CoalesceBatchesExec    │                                                          │   LiquidCacheClientExec   ││      RepartitionExec      │
                                                          │    --------------------   ││    --------------------   │                                                          │    --------------------   ││    --------------------   │
                                                          │          files: 1         ││     target_batch_size:    │                                                          │          server:          ││ partition_count(in->out): │
                                                          │      format: parquet      ││           16384           │                                                          │  http://localhost:50051,  ││           1 -> 8          │
                                                          │                           ││                           │                                                          │        mode=liquid,       ││                           │
                                                          │                           ││                           │                                                          │      object_stores=[]     ││    partitioning_scheme:   │
                                                          │                           ││                           │                                                          │                           ││     RoundRobinBatch(8)    │
                                                          └───────────────────────────┘└─────────────┬─────────────┘                                                          └─────────────┬─────────────┘└─────────────┬─────────────┘
                                                                                       ┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                                                                                       │        HashJoinExec       │                                                          │       DataSourceExec      ││       ProjectionExec      │
                                                                                       │    --------------------   │                                                          │    --------------------   ││    --------------------   │
                                                                                       │            on:            │                                                          │          files: 1         ││        ps_partkey:        │
                                                                                       │  (p_partkey = ps_partkey) │                                                          │      format: parquet      ││         ps_partkey        │
                                                                                       │                           │                                                          │                           ││                           │
                                                                                       │                           ├──────────────┐                                           │                           ││       ps_supplycost:      │
                                                                                       │                           │              │                                           │                           ││       ps_supplycost       │
                                                                                       │                           │              │                                           │                           ││                           │
                                                                                       │                           │              │                                           │                           ││        s_nationkey:       │
                                                                                       │                           │              │                                           │                           ││        s_nationkey        │
                                                                                       └─────────────┬─────────────┘              │                                           └───────────────────────────┘└─────────────┬─────────────┘
                                                                                       ┌─────────────┴─────────────┐┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐
                                                                                       │   LiquidCacheClientExec   ││   LiquidCacheClientExec   │                                                          │    CoalesceBatchesExec    │
                                                                                       │    --------------------   ││    --------------------   │                                                          │    --------------------   │
                                                                                       │          server:          ││          server:          │                                                          │     target_batch_size:    │
                                                                                       │  http://localhost:50051,  ││  http://localhost:50051,  │                                                          │           16384           │
                                                                                       │        mode=liquid,       ││        mode=liquid,       │                                                          │                           │
                                                                                       │      object_stores=[]     ││      object_stores=[]     │                                                          │                           │
                                                                                       └─────────────┬─────────────┘└─────────────┬─────────────┘                                                          └─────────────┬─────────────┘
                                                                                       ┌─────────────┴─────────────┐┌─────────────┴─────────────┐                                                          ┌─────────────┴─────────────┐
                                                                                       │       DataSourceExec      ││       DataSourceExec      │                                                          │        HashJoinExec       │
                                                                                       │    --------------------   ││    --------------------   │                                                          │    --------------------   │
                                                                                       │          files: 1         ││          files: 1         │                                                          │            on:            │
                                                                                       │      format: parquet      ││      format: parquet      │                                                          │  (s_suppkey = ps_suppkey) │
                                                                                       │                           ││                           │                                                          │                           ├──────────────┐
                                                                                       │         predicate:        ││                           │                                                          │                           │              │
                                                                                       │   p_size = 15 AND p_type  ││                           │                                                          │                           │              │
                                                                                       │         LIKE %BRASS       ││                           │                                                          │                           │              │
                                                                                       └───────────────────────────┘└───────────────────────────┘                                                          └─────────────┬─────────────┘              │
                                                                                                                                                                                                           ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                                                                                                                                                                                                           │   LiquidCacheClientExec   ││   LiquidCacheClientExec   │
                                                                                                                                                                                                           │    --------------------   ││    --------------------   │
                                                                                                                                                                                                           │          server:          ││          server:          │
                                                                                                                                                                                                           │  http://localhost:50051,  ││  http://localhost:50051,  │
                                                                                                                                                                                                           │        mode=liquid,       ││        mode=liquid,       │
                                                                                                                                                                                                           │      object_stores=[]     ││      object_stores=[]     │
                                                                                                                                                                                                           └─────────────┬─────────────┘└─────────────┬─────────────┘
                                                                                                                                                                                                           ┌─────────────┴─────────────┐┌─────────────┴─────────────┐
                                                                                                                                                                                                           │       DataSourceExec      ││       DataSourceExec      │
                                                                                                                                                                                                           │    --------------------   ││    --------------------   │
                                                                                                                                                                                                           │          files: 1         ││          files: 1         │
                                                                                                                                                                                                           │      format: parquet      ││      format: parquet      │
                                                                                                                                                                                                           └───────────────────────────┘└───────────────────────────┘
